### ISSUE TITLE
HAProxy Lua AppletHTTP.receive() hangs in 2.8, works in 2.7.

### ISSUE THREAD

**Issue by fatchan** [haproxy/haproxy#2180]
### Detailed Description of the Problem

Hi, in HAProxy 2.8 when calling AppletHTTP.receive() in a lua function registered with core.register_service, it hangs and the call never returns. It works in 2.7.

Here is a repo to demonstrate the issue:

https://github.com/fatchan/haproxy-lua-2.8-bug/

You can start it with `docker-compose up`. It will listen on 127.0.0.1:8080. The test.lua is very simple and just prints the body and echos it back to you.

The docker-compose.yml file uses the `docker:2.7` image by default.  If you do:
```
curl -vvv http://127.0.0.1:8080 --data "x=1&y=2"
```
You will see in the haproxy logs:
```
...
test-haproxy-1  | before
test-haproxy-1  | x=1&y=2
test-haproxy-1  | after
...
```

When you change the image in the docker-compose.yml to `haproxy:2.8` and repeat the test, you will see:
```
...
test-haproxy-1  | before

```

Am I doing something wrong?

Kind Regards.

### Expected Behavior

I would expect it to work as it did in 2.7, because I don't see any difference in the documentation for AppletHTTP.receive.

### Steps to Reproduce the Behavior

1. `docker-compose up`
2. curl -vvv http://127.0.0.1:8080 --data "x=1&y=2"
3. observe

change the image in docker-compose.yml to haproxy:2.8 instead hf haproxy:2.7
1. `docker-compose up`
2. curl -vvv http://127.0.0.1:8080 --data "x=1&y=2"
3. observe

### What is your configuration?

```haproxy
https://github.com/fatchan/haproxy-lua-2.8-bug/blob/master/haproxy.cfg
```

### Output of `haproxy -vv`

```plain
N/A
```

### Additional Information

A similar issue was reported in the past #1207

**Comment by capflam** [comment]
Indeed, I can reproduce the bug. Thanks for the report, I'm on it !

**Comment by capflam** [comment]
I got it ! It is a bug in the http receive functions (recv and getline). We check the wrong side to detect the end of the message. Here is the patch, I will push a fix.

```diff
diff --git a/src/hlua.c b/src/hlua.c
index 742c82448..ae0d412a4 100644
--- a/src/hlua.c
+++ b/src/hlua.c
@@ -5353,7 +5353,7 @@ __LJMP static int hlua_applet_http_getline_yield(lua_State *L, int status, lua_K
        /* The message was fully consumed and no more data are expected
         * (EOM flag set).
         */
-       if (htx_is_empty(htx) && (sc->flags & SC_FL_EOI))
+       if (htx_is_empty(htx) && (sc_opposite(sc)->flags & SC_FL_EOI))
                stop = 1;
 
        htx_to_buf(htx, &req->buf);
@@ -5445,7 +5445,7 @@ __LJMP static int hlua_applet_http_recv_yield(lua_State *L, int status, lua_KCon
        /* The message was fully consumed and no more data are expected
         * (EOM flag set).
         */
-       if (htx_is_empty(htx) && (sc->flags & SC_FL_EOI))
+       if (htx_is_empty(htx) && (sc_opposite(sc)->flags & SC_FL_EOI))
                len = 0;
 
        htx_to_buf(htx, &req->buf);

```

**Comment by capflam** [comment]
pushed !

**Comment by capflam** [comment]
Now backported to 2.8

**Comment by fatchan** [comment]
Excellent, thank you very much! 🙏🏻

**Comment by fatchan** [comment]
Hi @capflam can I inquire when a 2.8 build with the patch included might be available from  https://haproxy.debian.net/ repo? Are you the right person, or is there somewhere I can find information about the schedule for releases?

Kind regards.

**Comment by capflam** [comment]
We should emit a 2.8.1 first. The debian package will quickly follow. For now, there is no ETA for the 2.8.1. But the 2.8.0 was released a month ago, I guess we will emit a 2.8.1 soon.
